wikimedia/wikipedia-ios
# Variant-aware feed settings

This is work towards the Chinese variants feature https://phabricator.wikimedia.org/T195265
It is not the entire feature/ticket.

There are four commits in the PR. It may be easiest to review each commit in turn.

### Notes
The overall purpose of this PR is for the very flexible and intricate preferences for the explore feed to support language variants.

Currently, these preferences are stored keyed by site URL string. (i.e. https://zh.wikipedia.org). This PR changes this to be stored by language variant code (i.e. en, zh-hans). The language variant code is now being used to store preferred languages and uniquely identifies a language / language variant.

**Commit 1: Add wmf_contentLanguageCode computed property to NSURL/URL**
Mirrors the behavior of MWKLanguageLink to provide the content language code.
Should be useful in general, but in this case it reduces the amount of change necessary in the WMFExploreFeedContentController by allowing it to continue to operate on site URLs.
Added tests as well.

**Commit 2: Have MWKLanguageLink URL properties include languageVariantCode**
When siteURL or articleURL is called on a MWKLanguageLink instance, make sure the languageVariantCode is set on the returned URL.
Added test as well.

**Commit 3: Add variant property to Wikipedia model entities**
Add a new Wikipedia.xcdatamodel version that adds a _variant_ property to WMFContentGroup and WMFArticle. This includes adding the appropriate header and implementation file code for the new properties. The use of the name _variant_ as opposed to _contentLanguageCode_ is for consistency for the existing _variant_ property on the CacheItem entity in the Cache model.

Update URL computed properties to propagate language variant code. Note that methods that create the WMFArticle do not yet set the language variant code. This was to keep this PR focused on the changes needed for making the feed preferences location-aware. The added property in WMFArticle was to try to keep model migrations/versions to a minimum.

Addition of these properties is handled by Core Data lightweight migration and the persistent store coordinator is already setup to perform lightweight migrations if needed.

**Commit 4: Make explore feed preferences language-variant aware**
The bulk of this commit is using the wmf_contentLanguageCode as the language key in the preferences instead of the wmf_databaseKey (which is basically the siteURL as a string).

There is code added to update the preferences from the old keys to the new keys. _Note that this code is a bootstrapped placeholder until the eventual 'move to variants' initial migration is in place._

Finally, much of this commit are updating method and variable names to accurately reflect the use of contentLanguageCode.

**Commit 5 & 6: Test additions**
Add more to an existing test and add a comment to clarify a test.

### Test Steps
1. With language variants turned off (the current default) there should be no visible change in the behavior of explore feed settings or the app at all.

2. With language variants turned on in the WikipediaLookup.swift file, and language variants chosen as preferred languages, all explore settings should display and work correctly, even with multiple variants from the same language.

To test, I used Serbian, Cyrillic and Serbian, Latin separated by another language. Putting a different language between the variants makes it easier to tell that a new set of explore cards for a different variant is present.

NOTE: The *explore feed settings* will work. The explore feed itself should show cards based on the settings, but the variant fetched and displayed will not be correct yet.
